### PR TITLE
JENKINS-60482 Alternate Endpoint fix

### PR DESCRIPTION
**Pull request by troymohl**
This fixes a bug in the EC2 Alternate URL endpoint configuration reported in https://issues.jenkins-ci.org/browse/JENKINS-60482. Additionally, small bug fix to fix cases if an empty endpoint got passed to the AmazonEC2 client initialization.

~This PR has https://github.com/jenkinsci/ec2-plugin/pull/438 as a parent, as my configuration required that fix to be in place in order to make this fix.~
This has been fixed in master, so that PR was no longer needed.

### CODE REVIEW

**Comment by five-iron** [comment]
Any update on this PR? This is a nice QOL fix. Looks like it's just pending review from @res0nance.

File: src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java:63]: Is the endpoint not required for the AmazonEC2 Client object? I don't see it specified for the object used when spinning up instances etc.

File: src/main/java/hudson/plugins/ec2/util/AmazonEC2FactoryImpl.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/util/AmazonEC2FactoryImpl.java:17]: Method should be annotated with CheckForNull

File: src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java
Comment (troymohl) [reply src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java:63]: I didn't change any of the existing code that checks for endpoint. I will look at switching the getEc2EndpointUrl logic to better handle where it's directly set and not use the region if the endpoint URL is configured. I think I did see a ticket where the region -> endpoint logic didn't work in C2S, but I won't have the ability to fully test that.

File: src/main/java/hudson/plugins/ec2/util/AmazonEC2FactoryImpl.java
Comment (troymohl) [reply src/main/java/hudson/plugins/ec2/util/AmazonEC2FactoryImpl.java:19]: Added annotation

**Comment by jkfernan** [comment]
Any update on this PR?

**Comment by troymohl** [comment]
Update this PR to use the Alternate EC2 Endpoint if configured. I believe this change would also fix using in C2S where the endpoint is not easily derived from the region. I am unable to test this though.

**Comment by GaToRAiD** [comment]
Seems like a simple PR to merge, this headache is intense when going into the settings to verify things and having to set the endpoint every time.

**Comment by GaToRAiD** [comment]
What is the status of this PR? It is quite a pain going into the configuration settings.

**Comment by GaToRAiD** [comment]
@troymohl what is the status on this?

**Comment by troymohl** [comment]
@GaToRAiD 
I have tested this on instances that I have access to, as well as rebasing against the master branch. I'm just an outside contributor that needed this bug fixed and wanted to contribute this back for others. I don't have any additional info on when/if this will be accepted into the main baseline.

**Comment by jkfernan** [comment]
@res0nance Can we get a code review on this? It would be great to get it merged. Thanks.

**Comment by alecharp** [comment]
@troymohl  can you have a look at the merge conflict? Thanks.

**Comment by GaToRAiD** [comment]
Any update on this PR?

**Comment by rsandell** [comment]
@res0nance @MRamonLeon I am confused about the status of this? @GaToRAiD mentions that it is already fixed in master, but I can't see the ticket mentioned in any release notes and I have users reporting it still being an issue?

**Comment by MRamonLeon** [comment]
> @res0nance @MRamonLeon I am confused about the status of this? @GaToRAiD mentions that it is already fixed in master, but I can't see the ticket mentioned in any release notes and I have users reporting it still being an issue?

Maybe because of that: https://github.com/jenkinsci/ec2-plugin/pull/330 Did you mean this @GaToRAiD?

**Comment by MRamonLeon** [comment]
> Disregard, looks like it was already fixed in master.

Could you tell us where it was fixed please @GaToRAiD? Thanks

**Comment by mrmateo** [comment]
@MRamonLeon / @rsandell - I think @GaToRAiD might have been referring to an edit in the original comment on this PR from @troymohl :
> This has been fixed in master, so that PR was no longer needed.

If so, his comment there was only referring to a prerequisite fix that was needed for this and not the actual issue that this PR fixes. That prerequisite fix was merged and is in master.

@rsandell / @alecharp / @res0nance - Can one of you please take another look and either approve or deny this? It is impacting many users of this plugin so would be good to know if we can count on this fix being made soon or if we need to find other solutions.

Thanks!

**Comment by res0nance** [comment]
@mrmateo  This is unfortunately impossible to review due to the conflicts

**Comment by troymohl** [comment]
I have rebased this against the latest master.

**Comment by batmat** [comment]
@mrmateo please do not hesitate to help review this, we welcome any help! :-)

**Review by res0nance — APPROVED** [review]

**Review by MRamonLeon — CHANGES_REQUESTED** [review]
Looking into it, please don't merge yet (despite the conflicts)

**Comment by MRamonLeon** [comment]
@troymohl I've fixed the conflicts with master, but after reviewing the PR, I find we maybe go too far and an simpler approach can be used, if I understood your problem from your description in the Jira ticket. I've filed https://github.com/jenkinsci/ec2-plugin/pull/537 with this approach where we just ensure the _Alternate EC2 endpoint_ is persisted. 

I've recorded a video to show how is the behavior, just in case you or someone else have some confusion about how the configuration page works (it's deserving a good refactoring, let's be honest :wink: )

Let me know if this approach fits your needs.

cc: @jkfernan @GaToRAiD

**Comment by troymohl** [comment]
@MRamonLeon This actually started with a similar change: https://github.com/jenkinsci/ec2-plugin/pull/439/commits/d200d115ccd1e3a277021ba1d170ec0cc8521921

Based on: https://github.com/jenkinsci/ec2-plugin/pull/439#discussion_r412273294 I made the additional changes. I'm good with whatever the team decides. We can close this PR if the approach in https://github.com/jenkinsci/ec2-plugin/pull/537 is preferred.

**Comment by MRamonLeon** [comment]
> @MRamonLeon This actually started with a similar change: [d200d11](https://github.com/jenkinsci/ec2-plugin/commit/d200d115ccd1e3a277021ba1d170ec0cc8521921)
> 
> Based on: [#439 (comment)](https://github.com/jenkinsci/ec2-plugin/pull/439#discussion_r412273294) I made the additional changes. I'm good with whatever the team decides. We can close this PR if the approach in #537 is preferred.

@troymohl I'm happy merging this PR instead of mine if you come back to your original commit, https://github.com/jenkinsci/ec2-plugin/commit/d200d115ccd1e3a277021ba1d170ec0cc8521921

**Comment by MRamonLeon** [comment]
#537 merged. Closing this one.

**Comment by MRamonLeon** [comment]
@troymohl thank you for filing this PR and leveraging the fix of this issue. I look forward to seeing more contributions from you. :clap:

**Pull request closed without merging**
